app/Main.hs

```diff
@@ -237,17 +239,66 @@ getHttpManager opts = newManager managerSettings
 --
 -- Signals failure through a value of type VaultError, but can also
 -- throw HTTP exceptions.
-vaultEnv :: Context -> ExceptT VaultError IO [EnvVar]
+vaultEnv :: Context -> IO (Either VaultError [EnvVar])
 vaultEnv context = do
-  mountInfo <- requestMountInfo context
-  secrets <- mapExceptT (fmap $ first SecretFileError) $ readSecretList secretFile
-  secretEnv <- requestSecrets context mountInfo secrets
-  checkNoDuplicates (buildEnv secretEnv)
+  mountInfo <- doWithRetries getMountInfo
+  case mountInfo of
+    Left vaultError -> return $ Left vaultError
+    Right mountInfo' -> do
+      secrets <- readSecretsFile secretFile
+      case secrets of
+        Left sfError -> return $ Left $ SecretFileError sfError
+        Right secrets' -> do
+          -- NOTE: This retries fetching all the secrets if fetching one of the
+          -- secrets fails due to an error we should retry on. We would ideally
+          -- keep the secrets that were successfully fetched in memory and only
+          -- retry getting the ones we couldn't get yet, as this makes it less
+          -- likely to get a 'thundering herd' of Vaultenv processes all trying
+          -- and failing to get all of their secrets.
+          -- This would then need to happen in requestSecrets, but I don't want
+          -- to touch that right now.
```

Comment:
Let's file a new issue for this.

Reply:
Done, #90. Link added to this comment as well.